SEP-12 callback endpoint (#805)
stellar/stellar-protocol: ecosystem/sep-0012.md

> Author: Interstellar

Reviewer: 💡If you'd like you can add your name to the list of authors here.

Status: Active
Created: 2018-09-11
Updated: 2020-12-23

> Version 1.2.0

Reviewer: ❗This change adds a new feature can you bump the minor version?

```suggestion
 Version 1.3.0
```

This SEP was made with these goals in mind:

> - interoperability

Reviewer: ❗This is an active SEP. It's helpful if we keep changes surgical, so it's ultra clear to implementers what has changed. We can refactor but like code it's helpful if the refactor commits make no functional change so implementers can ignore them. can you move the format changes to a separate PR? I'd also question if the format changes are necessary. Even with refactors we should weigh the return on investment vs churning diffs and making it harder to follow the diffs history.

Author: Sorry, the reformatting happened automatically. just pushed it with the formatting removed.

- give customers control over their data by supporting complete data erasure

To support this protocol an anchor acts as a server and implements the specified REST API endpoints, while a wallet implements a client that consumes the API. The goal is interoperability, so a wallet implements a single client according to the protocol, and will be able to interact with any compliant anchor. Similarly, an anchor that implements the API endpoints according to the protocol will work with any compliant wallet.

## Prerequisites

- An anchor must define the location of their `KYC_SERVER` or `TRANSFER_SERVER` in their [`stellar.toml`](sep-0001.md). This is how a client app knows where to find the anchor's server. A client app will send KYC requests to the `KYC_SERVER` if it is specified, otherwise to the `TRANSFER_SERVER`.
- Anchors and clients must support [SEP-10](sep-0010.md) web authentication and use it for all SEP-12 endpoints.

## API Endpoints

- [`GET /customer`](#customer-get): Check the status of a customers info
- [`PUT /customer`](#customer-put): Idempotent upload of customer info
- [`DELETE /customer`](#customer-delete): Deletion of customer info

> - [`POST /customer`](#customer-post): Set callbacks for a wallet to receive status updates from the anchor

Reviewer: ❗I think this breaks the existing REST  pattern this SEP exposes. The description indicates that this sets a value which implies to me it can be called with an updated value. For consistency with the rest of this SEP it should use the `PUT`.

❓Is there a need to set the callback in a separate call? I think the new `on_change_callback` field should be added added to the existing `PUT /customer`. If there's a need to set the callback separately then we should make it a sub-path, e.g. `PUT /customer/callbacks`.

Author: I think the POST verb applies better than the PUT verb. PUT implies idempotency, but a wallet may want to set more than one callback on a given KYC user.

I don't feel strongly about `/customer` vs. `/customer/callbacks`, and I think the distinction is a semantic / aesthetic one, so I'll just take your suggestion and change the URL.

Reviewer: Is the reason there would be multiple because an account holder might be using several different wallet applications, and each might register for callbacks?

I'd expect if it's just one application it'd be one callback url and if it a new url is generated updating would be sufficient, and so PUT would be more appropriate.

Could you elaborate on the multiple use case?

If we have multiple we need to think about deleting or unregistering callbacks. That probably means returning a callback id. Or define when they expire naturally. Exposing might be better so that old wallets don't get callbacks after a user switches wallets.

If we require the callback to be included in the main PUT call we could tie cannabis to that specific kyc run, as opposed to being registered with the account forever. Once that kyc run completes it could be considered expired.

Reviewer: Is the reason there would be multiple because an account holder might be using several different wallet applications, and each might register for callbacks?

I'd expect if it's just one application it'd be one callback url and if it a new url is generated updating would be sufficient, and so PUT would be more appropriate.

Could you elaborate on the multiple use case?

If we have multiple we need to think about deleting or unregistering callbacks. That probably means returning a callback id. Or define when they expire naturally. Exposing might be better so that old wallets don't get callbacks after a user switches wallets.

If we require the callback to be included in the main PUT call we could tie cannabis to that specific kyc run, as opposed to being registered with the account forever. Once that kyc run completes it could be considered expired.

Reviewer: Returning to a question I asked that wasn't answered:
>Is the reason there would be multiple because an account holder might be using several different wallet applications, and each might register for callbacks?

I think this is a valid use case, and so it would have also been fine to support multiple callbacks but clean up the REST pattern in use to support registering a callback and deleting it. Example:
- `POST /customer/callbacks` returns a callback ID
- `DELETE /customer/callbacks/<id>` deletes the callback (optional given that callbacks stop once the account arrives at a terminal state).

If a single callback is all we need in the foreseeable future then we don't need to add the burden of multiple callbacks to SEP-12 implementations, and we can always add other endpoints when they're needed at a later date.

Success | `200 OK`
Client not authenticated properly | `401 Unauthorized`
Anchor has no information on the customer | `404 Not Found`

## Customer callback POST

Allow the wallet to provide a callback URL to the anchor.

Whenever the user's
`status` field changes, the anchor will issue a POST request to the callback
URL. The payload of the POST request will be the same as [`GET /customer`](#customer-get).

If a wallet submits a callback URL that has already been set, the anchor shouldn't POST changes more than once. But a wallet might submit more than one unique callback URL.

### Request

```
POST [KYC_SERVER || TRANSFER_SERVER]/customer/callbacks
```

| Name                 | Type     | Description             |
| -------------------- | -------- | ----------------------- |
| `id`        | `string`      | (optional) The ID of the customer as returned in the response of a previous `PUT` request. If the customer has not been registered, they do not yet have an `id`.                                                                                                                                     |

> | `account`   | `G...` string | (optional) The Stellar account ID used to identify this customer. If many customers share the same Stellar account, the `memo` and `memo_type` parameters should be included as well.                                                                                                                 |

Reviewer: Can you explicitly list the memo fields?

Success | `200 OK`
Client not authenticated properly | `401 Unauthorized`
Anchor has no information on the customer | `404 Not Found`

## Customer callback PUT

Allow the wallet to provide a callback URL to the anchor. The provided callback URL will replace (and supercede) any previously-set callback URL for this account.

Whenever the user's `status` field changes, the anchor will issue a POST request to the callback URL. The payload of the POST request will be the same as [`GET /customer`](#customer-get).

> It's the wallet's responsibility to send a working callback URL to the anchor. If callback POST fails due to a problem on the callback URL's side, the anchor should not retry it.

Reviewer: >If callback POST fails due to a problem on the callback URL's side, the anchor should not retry it.

❗I think we should allow retrying, but leave the specifics undefined. Retrying callbacks is a common practice, but we can't be too prescriptive without burdening services that don't have the capabilities to support robust callback retry systems.

By disallowing retries here we encourage clients to build systems that can't handle duplicate requests, which could be dangerous. A callback URL should always handle cases like being called more than once, or other edge cases, and there's little reason to disallow retries when a callback endpoint needs to handle those other cases.

```suggestion
It's the wallet's responsibility to send a working callback URL to the anchor. If callback POST fails due to a problem on the callback URL's side, the anchor may retry it, although is not required to.
```

Alternatively, if we don't feel like we want to make a decision on this just yet, delete this statement and leave retries undefined until such time as they become defined.

```suggestion
It's the wallet's responsibility to send a working callback URL to the anchor.
```

### Request

```
PUT [KYC_SERVER || TRANSFER_SERVER]/customer/callback
```

| Name                 | Type     | Description             |
| -------------------- | -------- | ----------------------- |
| `id`        | `string`      | (optional) The ID of the customer as returned in the response of a previous `PUT` request. If the customer has not been registered, they do not yet have an `id`.                                                                                                                                     |
| `account`   | `G...` string | (optional) The Stellar account ID used to identify this customer. If many customers share the same Stellar account, the `memo` and `memo_type` parameters should be included as well.                                                                                                                 |
`memo` | `string` | (optional) The memo used to create the customer record |
`memo_type` | `string` | (optional) The type of memo used to create the customer record |

> | `url` | `string` | A valid URL. |

Reviewer: 💡 Could we include some more detail here about what this URL is? Example:

```suggestion
| `url` | `string` | A callback URL that the SEP-12 server will POST to when the state of the account changes. |
```

Success | `200 OK`
Client not authenticated properly | `401 Unauthorized`
Anchor has no information on the customer | `404 Not Found`

> ## Customer callback PUT

Reviewer 2: Can we remove the newline characters in this section to match the style of the rest of the doc?

Allow the wallet to provide a callback URL to the anchor. The provided callback 
URL will replace (and supercede) any previously-set callback URL for this account.

Whenever the user's
`status` field changes, the anchor will issue a POST request to the callback
URL. The payload of the POST request will be the same as [`GET /customer`](#customer-get).

Anchors will submit POST requests until the user's status changes to `ACCEPTED`
or `REJECTED`. If a wallet needs to watch a user's KYC status after that,
it will need to set a callback again.

> ### Request

Reviewer 2: If a callback request to the provided URL fails, is the server obligated to retry? I don't think it is, but we should make that clear in this section.

| -------------------- | -------- | ----------------------- |
| `id`        | `string`      | (optional) The ID of the customer as returned in the response of a previous `PUT` request. If the customer has not been registered, they do not yet have an `id`.                                                                                                                                     |
| `account`   | `G...` string | (optional) The Stellar account ID used to identify this customer. If many customers share the same Stellar account, the `memo` and `memo_type` parameters should be included as well.                                                                                                                 |
`memo` | `string` | (optional) The memo used to create the customer record |
`memo_type` | `string` | (optional) The type of memo used to create the customer record |
| `url` | `string` | A valid URL. |

### PUT Responses

| Situation                                 | Response           |
| ----------------------------------------- | ------------------ |
| Success                                   | `200 OK`           |

> | Client not authenticated properly         | `401 Unauthorized` |

Reviewer 2: I think we should stick to using `403 Permission Denied` since thats used in SEP 6, 24, & 31. Maybe others as well.

Reviewer 2: We should probably add a comment about using 403 in the authentication section since it applies to any request here.

Author: The other, existing calls already return 401. Are you sure you want to change the error response if this SEP is already in use?

Reviewer 2: Ah you're right, darn. I don't want to make any breaking changes so lets just stick with 401.
